### PR TITLE
[D2M] DMA Region Merging

### PR DESCRIPTION
**Pull request by bgrady-tt**
### Ticket
See https://github.com/tenstorrent/tt-mlir/issues/4621

### Problem description
In order to have fully streaming binary ops, we must merge >2 DMA regions onto 2 physical DMA regions we can implement on a BRISC/NCRISC pair.

### What's changed
- GenericHWThreadSelection has swapped positions in pipeline with GenericLowerDMAs
  - DMA threads are merged _before_ lowering
- GenericHWThreadSelection merges all DMA regions according to following scheme
  - First, input DMA regions are round-robin merged into DMA regions 0 and 1 to balance load
  - Second, output DMA regions are always merged to ^datamovement1, to ensure they have access to NOC1
    - **rationale**: Most streaming DMA writes will be to DRAM, where using NOC1 is very important to avoid congestion. Even if this results in two output threads sharing NOC1, that is still generally preferable than using NOC0 for DRAM writes.
  - For correctness, output DMA threads are always merged _after_ all input DMA blocks within each DMA region
  - **special case for local ttir.await**: merged into _compute region_ to avoid blocking dma threads unnecessarily (same behavior as before). 
    - If ^datamovement1 is not used for input DMA, the local ttir.await is left in ^datamovement1
  - If a compute region is **NOT** present in original op:
    - DMA Merging is still performed, but all excess DMA threads are merged to datamovement regions
    - A new compute region is NOT created

### Checklist
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: test/ttmlir/Dialect/TTIR/generic/generic_hw_thread_selection.mlir [hunk 1]
```diff
@@ -1,4 +1,5 @@
-// RUN: ttmlir-opt --ttcore-register-device --ttir-generic-hw-thread-selection %s
+// RUN: ttmlir-opt --ttcore-register-device --ttir-generic-hw-thread-selection -o %t %s
```
Comment (bgrady-tt) [inline test/ttmlir/Dialect/TTIR/generic/generic_hw_thread_selection.mlir:1]: **note**: this mlir file had `CHECK` statements, but was not being piped to `FileCheck` at test time! 

So some unrelated changes were made as original `CHECK` statements actually didn't pass _even on origin/main_.

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp [hunk 2]
```diff
@@ -37,25 +39,32 @@ class TTIRGenericMoveTrivialOutputThreadToComputeRewritePattern
...
 
   LogicalResult matchAndRewrite(GenericOp op,
                                 PatternRewriter &rewriter) const final {
     unsigned outputOperandsIndex = op.getOutputs().getBeginOperandIndex();
+    unsigned inputOperandsLength = op.getInputs().size();
     unsigned outputOperandsLength = op.getOutputs().size();
-    assert(outputOperandsLength == 1);
-    if (outputOperandsIndex >= op.getNumRegions()) {
+    assert(inputOperandsLength <= 2 &&
+           "only 2 or less input operands are supported");
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: I think this will fall over as soon as we land fusion which is in the works.

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp [hunk 3]
```diff
@@ -37,25 +39,32 @@ class TTIRGenericMoveTrivialOutputThreadToComputeRewritePattern
...
+    assert(outputOperandsLength == 1 && "only one output operand is supported");
+    // note: If only one input + one output operand, dma is already in optimal
+    // form
+    if (inputOperandsLength < 2 || outputOperandsIndex >= op.getNumRegions()) {
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: Maybe we can replace this check with the same assertion we have below:
```
      if (op.getNumRegions() <= (chipDesc.getNumComputeThreads() +
                                chipDesc.getNumDatamovementThreads())) {
```

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp [hunk 4]
```diff
@@ -37,25 +39,32 @@ class TTIRGenericMoveTrivialOutputThreadToComputeRewritePattern
...
       return failure();
     }
     Region &outputRegion = op.getRegion(outputOperandsIndex);
-    Block *outputBlock = getIfTrivialBlock(outputRegion, outputOperandsIndex);
-    if (!outputBlock) {
+    if (outputRegion.getBlocks().size() == 0) {
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: I think this should be an assertion, or rather the assertion below should handle this case.  We shouldn't support a generic regions without a block.  Is there an edge case that I'm missing?

File: test/ttmlir/Dialect/TTIR/generic/generic_hw_thread_selection.mlir (hunk 1 above)
Comment (nsmithtt) [reply test/ttmlir/Dialect/TTIR/generic/generic_hw_thread_selection.mlir:1]: Good catch!

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp [hunk 5]
```diff
@@ -75,8 +84,19 @@ class TTIRGenericMoveTrivialOutputThreadToComputeRewritePattern
...
+    // Output DMA regions that contain a lone ttir.await are special, in that
+    // they can (and should) be merged into compute region, were they are later
+    // transformed into a local cb_wait_front()/cb_pop_front() pair with no DMA.
+    bool local_cb_pop_only =
+        isLocalAwaitForOutputCB(outputRegion, outputOperandsIndex);
+    unsigned mergeRegionIndex =
+        (local_cb_pop_only) ? computeRegionIndex : lastInputRegionIndex;
+
+    Block *newBlock = &newGeneric.getRegion(mergeRegionIndex).back();
+    rewriter.mergeBlocks(&outputBlock, newBlock, newBlock->getArguments());
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: How does this work in conjunction with the code above?  The code from lines 71-85 unconditionally erase the output datamovement thread, but we're no longer really checking if that needs to happen.

IMO we should restructure this rewriter to work recursively:
- Do we exceed max hw threads? if so, continue, else failure because we're done
- identify best thread to be merged (i.e. a trivial await block would be best candidate first)
- identify best thread to merge into
- merge + erase that region from the generic op
- let the greedy rewrite driver recursively apply

LMK what you think, the above could be adapted to one shot the whole thing if we think we'll lose information between recursive applications.

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp (hunk 3 above)
Comment (bgrady-tt) [reply lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: Sure I can make this change

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp (hunk 5 above)
Comment (bgrady-tt) [reply lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: I think collapsing this in one shot is a better approach, since recursive updates just make the analysis harder (i.e. each time you need to re-analyze each merged region and decide best merge candidate)

Comment (bgrady-tt) [reply lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: I can make this merging work for arbitrary numbers of input DMA threads. A reasonable outcome IMO is to prefer ^datamovement0 for input threads and ^datamovement1 for output threads, but otherwise try to keep DMA load balanced between the two regions.

A followup PR could explore heuristics for analyzing relative DMA intensity of each thread and attempt to balance the load optimally.

Comment (nsmithtt) [reply lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: Sounds good, though it's still probably worth incrementally choose merge candidates, if you consider a larger example with regions:
- `A B C D E F G`

The best merge strategy could result in 3 regions (datamovement, datamovement, compute):
- `AC BDE GF`

Comment (nsmithtt) [reply lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: > ^datamovement0 for input threads and ^datamovement1 for output threads, but otherwise try to keep DMA load balanced between the two regions.
>
> A followup PR could explore heuristics for analyzing relative DMA intensity of each thread and attempt to balance the load optimally.

Makes sense, sounds great!

Comment (bgrady-tt) [reply lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: @nsmithtt updated the logic and tests to allow merging arbitrary numbers of inputs and output DMA regions. The only early exit special case is now `kInputs + kOutputs <= 1`.

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp [hunk 6]
```diff
@@ -37,46 +39,100 @@ class TTIRGenericMoveTrivialOutputThreadToComputeRewritePattern
...
+    assert(chipIds.size() == 1);
+    auto chipDesc = systemDesc.getChipDesc(chipIds[0]);
+
     unsigned outputOperandsIndex = op.getOutputs().getBeginOperandIndex();
-    unsigned outputOperandsLength = op.getOutputs().size();
-    assert(outputOperandsLength == 1);
-    if (outputOperandsIndex >= op.getNumRegions()) {
+    unsigned kInputs = op.getInputs().size();
+    unsigned kOutputs = op.getOutputs().size();
+    unsigned kDMAHWThreads = chipDesc.getNumDatamovementThreads();
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: Our style guide reserves the `k` prefix for enums / static constants. Can we replace with `num` perhaps?

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp [hunk 7]
```diff
@@ -37,46 +39,100 @@ class TTIRGenericMoveTrivialOutputThreadToComputeRewritePattern
...
+
+    // handling special case of single datamovement region is very clumsy and
+    // achieves nothing, so skip it
+    if (kInputs + kOutputs < 2) {
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: Shouldn't this read `if (kInputs + kOutputs < kDMAHWThreads)`?

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp [hunk 8]
```diff
@@ -37,46 +39,100 @@ class TTIRGenericMoveTrivialOutputThreadToComputeRewritePattern
...
     unsigned outputOperandsIndex = op.getOutputs().getBeginOperandIndex();
-    unsigned outputOperandsLength = op.getOutputs().size();
-    assert(outputOperandsLength == 1);
-    if (outputOperandsIndex >= op.getNumRegions()) {
+    unsigned kInputs = op.getInputs().size();
+    unsigned kOutputs = op.getOutputs().size();
+    unsigned kDMAHWThreads = chipDesc.getNumDatamovementThreads();
+
+    // handling special case of single datamovement region is very clumsy and
+    // achieves nothing, so skip it
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: Technically we require all comments to start titlecase and end with a `.`.  Not a fan of the rule personally, but it's in our style guide.

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp [hunk 9]
```diff
@@ -37,46 +39,100 @@ class TTIRGenericMoveTrivialOutputThreadToComputeRewritePattern
...
+    if (kInputs + kOutputs < 2) {
       return failure();
     }
-    Region &outputRegion = op.getRegion(outputOperandsIndex);
-    Block *outputBlock = getIfTrivialBlock(outputRegion, outputOperandsIndex);
-    if (!outputBlock) {
+
+    if (outputOperandsIndex >= op.getNumRegions()) {
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: I know this condition was here before, but I'm struggling to see what case this handles.

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp [hunk 10]
```diff
@@ -37,46 +39,100 @@ class TTIRGenericMoveTrivialOutputThreadToComputeRewritePattern
...
       return failure();
     }
 
-    SmallVector<Attribute> threads(op.getThreads().getValue());
-    // Skip the output operands block
-    threads.erase(threads.begin() + outputOperandsIndex);
+    llvm::SmallDenseSet<unsigned> mergedIndices;
+    for (unsigned i = kDMAHWThreads; i <= outputOperandsIndex; ++i) {
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: This seems like it'd read clearer to me as `i < numDatamovementRegions`. ignore this comment if you go with the one just below.

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp [hunk 11]
```diff
@@ -37,46 +39,100 @@ class TTIRGenericMoveTrivialOutputThreadToComputeRewritePattern
...
+      mergedIndices.insert(i);
+    }
+
+    SmallVector<Attribute> threads;
+    for (unsigned idx = 0; idx < op.getThreads().size(); ++idx) {
+      if (!mergedIndices.contains(idx)) {
+        threads.push_back(op.getThreads()[idx]);
+      }
+    }
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: Maybe the above 2 for loops could be made more succinct:
```
SmallVector<Attribute> threads = op.getThreads().take_front(kDMAHWThreads);
```

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp [hunk 12]
```diff
@@ -37,46 +39,100 @@ class TTIRGenericMoveTrivialOutputThreadToComputeRewritePattern
...
+        dmaInputThreadsMerged++;
       }
-      rewriter.modifyOpInPlace(
-          op, [&] { newGeneric.getRegion(regionIndex++).takeBody(region); });
+      Block *mergeDestBlock =
+          &newGeneric.getRegion(dmaMergeTargetIndex).front();
+
+      rewriter.mergeBlocks(mergeSrcBlock, mergeDestBlock,
+                           mergeDestBlock->getArguments());
     }
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp:136]: This looks great!

**Review by nsmithtt — COMMENTED** [review]
Mostly cosmetic, feedback inline, but overall this looks great!

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp (hunk 7 above)
Comment (bgrady-tt) [reply lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: I can just skip the rewrite if `kInputs + kOutputs == kDMAHWThreads`. 

Fwiw It still works as-is; the following code basically copies all regions verbatim and then performs 0 merges.

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp (hunk 8 above)
Comment (bgrady-tt) [reply lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: Will Fix 👍

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp (hunk 9 above)
Comment (bgrady-tt) [reply lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: I believe this has been retained from original version of `GenericHWThreadSelection.cpp`, it seems incoherent/impossible to me. I will remove it.

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp (hunk 11 above)
Comment (bgrady-tt) [reply lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: Point taken that this is pretty ugly. I can probably just generate a new SmallVector of thread attributes de novo with kDMAHWThreads datamovement threads and 1 compute thread.

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp (hunk 7 above)
Comment (nsmithtt) [reply lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: I meant to write `<=` above, it should be `<=` right?  We'd want to skip if you only had 1 datamovement region for e.g.

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp [hunk 13]
```diff
@@ -37,46 +39,91 @@ class TTIRGenericMoveTrivialOutputThreadToComputeRewritePattern
...
 
-    SmallVector<Attribute> threads(op.getThreads().getValue());
-    // Skip the output operands block
-    threads.erase(threads.begin() + outputOperandsIndex);
+    // The final merged region will always have numDMAHWThreads datamovement
+    // threads and 1 compute thread.
+    SmallVector<Attribute> threads(
+        numDMAHWThreads,
+        rewriter.getAttr<ThreadAttr>(ThreadType::Datamovement));
+    threads.push_back(rewriter.getAttr<ThreadAttr>(ThreadType::Compute));
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: Maybe we should assert that the last region is in fact a compute? There are certainly dm only generic's being used today, albeit they're single threaded.

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp [hunk 14]
```diff
@@ -37,46 +39,91 @@ class TTIRGenericMoveTrivialOutputThreadToComputeRewritePattern
...
   }
 
   LogicalResult matchAndRewrite(GenericOp op,
                                 PatternRewriter &rewriter) const final {
+
+    auto device = ttcore::lookupDevice(op);
+    auto systemDesc = ttcore::getCurrentScopeSystemDesc(op);
+    auto chipIds = device.getChipIds();
+    assert(chipIds.size() == 1);
+    auto chipDesc = systemDesc.getChipDesc(chipIds[0]);
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: Maybe let's do this at the pass level and just pass in `chipDesc.getNumDatamovementThreads()` to this pattern rewriter's constructor.

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp (hunk 13 above)
Comment (bgrady-tt) [reply lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: I am adjusting this to allow DMA thread merging for ops without compute regions. Seems likely that a concat style op could be 3+ dma threads without any compute.

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp (hunk 14 above)
Comment (bgrady-tt) [reply lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: sure, I will move this up to reduce overhead

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp (hunk 13 above)
Comment (bgrady-tt) [reply lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: See changes in commit https://github.com/tenstorrent/tt-mlir/pull/4651/commits/bdea8163b78afd3e79e8ced8e015f1039770493d

File: lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp (hunk 14 above)
Comment (bgrady-tt) [reply lib/Dialect/TTIR/Transforms/GenericHWThreadSelection.cpp]: See https://github.com/tenstorrent/tt-mlir/pull/4651/commits/bdea8163b78afd3e79e8ced8e015f1039770493d
